vpr/src/place/initial_placement.cpp

```diff
@@ -236,6 +257,191 @@ static bool is_block_placed(ClusterBlockId blk_id) {
+                        legal = true;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    return legal;
+}
+
+static bool find_centroid_neighbor(t_pl_loc centroid_loc, t_logical_block_type_ptr block_type, int& cx_to, int& cy_to) {
+    const auto& compressed_block_grid = g_vpr_ctx.placement().compressed_block_grids[block_type->index];
+
+    //Determine centroid location in the compressed space of the current block
+    int cx_centroid = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_x, centroid_loc.x);
+    int cy_centroid = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_y, centroid_loc.y);
+
+    //Determine rlim in each direction
+    int first_rlim = 15;
+    int rlim_x = std::min<int>(compressed_block_grid.compressed_to_grid_x.size(), first_rlim);
+    int rlim_y = std::min<int>(compressed_block_grid.compressed_to_grid_x.size(), first_rlim);
+
+    //Determine the valid compressed grid location ranges
+    int min_cx, max_cx, delta_cx;
+    int min_cy, max_cy;
+
+    min_cx = std::max(0, cx_centroid - rlim_x);
+    max_cx = std::min<int>(compressed_block_grid.compressed_to_grid_x.size() - 1, cx_centroid + rlim_x);
+
+    min_cy = std::max(0, cy_centroid - rlim_y);
+    max_cy = std::min<int>(compressed_block_grid.compressed_to_grid_y.size() - 1, cy_centroid + rlim_y);
+
+    delta_cx = max_cx - min_cx;
+    bool legal = false;
+    int cx_from = -1;
+    int cy_from = -1;
+    legal = find_compatible_compressed_loc_in_range(block_type, min_cx, max_cx, min_cy, max_cy, delta_cx, cx_from, cy_from, cx_to, cy_to, false);
+
+    return legal;
+}
+
+static std::vector<ClusterBlockId> find_centroid_loc(t_pl_macro pl_macro, t_pl_loc& centroid) {
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    int x, y;
+    float acc_weight = 0;
+    float acc_x = 0;
+    float acc_y = 0;
+
+    ClusterBlockId b_from = pl_macro.members.at(0).blk_index;
+    std::vector<ClusterBlockId> connected_blocks_to_update;
+
+    //iterate over the from block pins
+    for (ClusterPinId pin_id : cluster_ctx.clb_nlist.block_pins(b_from)) {
+        ClusterNetId net_id = cluster_ctx.clb_nlist.pin_net(pin_id);
+
+        /* Ignore the special case nets which only connects a block to itself  *
+         * Experimentally, it was found that this case greatly degrade QoR     */
+        if (cluster_ctx.clb_nlist.net_sinks(net_id).size() == 1) {
+            ClusterBlockId source = cluster_ctx.clb_nlist.net_driver_block(net_id);
+            ClusterPinId sink_pin = *cluster_ctx.clb_nlist.net_sinks(net_id).begin();
+            ClusterBlockId sink = cluster_ctx.clb_nlist.pin_block(sink_pin);
+            if (sink == source) {
+                continue;
+            }
+        }
+
+        //if the pin is driver iterate over all the sinks
+        if (cluster_ctx.clb_nlist.pin_type(pin_id) == PinType::DRIVER) {
+            //continue on ignored nets
```

Comment:
Delete comment or go to a higher level (ignored nets are globally routed clocks etc; can skip them).